File: backends/nest/src/adapter/postgres.ts
// Postgres adapter over node-postgres (pg). The interesting method is
// SectorsRepo.decrementInventory: a single conditional UPDATE that makes overselling
// impossible at the database, exactly as in the Go and FastAPI backends.

import { Pool } from 'pg';
import { validate as isUuid } from 'uuid';

import { Errors } from '../domain/errors';
import {
  Event,
  EventStatus,
  Order,
  OrderStatus,
  Payment,
  PaymentStatus,
  QueueStatus,
  QueueToken,
  Reservation,
  ReservationStatus,
  Role,
  Sector,
  User,
} from '../domain/models';
import {
  EventRepository,
  OrderRepository,
  PaymentRepository,
  QueueRepository,
  ReservationRepository,
  SectorRepository,
  UserRepository,
} from '../usecase/ports';

const UNIQUE_VIOLATION = '23505';

function isUniqueError(err: unknown): boolean {
  return typeof err === 'object' && err !== null && (err as { code?: string }).code === UNIQUE_VIOLATION;
}

export class UsersRepo implements UserRepository {
  constructor(private pool: Pool) {}
  private row = (r: any): User => ({
    id: r.id,
    email: r.email,
    passwordHash: r.password_hash,
    role: r.role as Role,
    createdAt: r.created_at,
  });
  async findByEmail(email: string) {
    const { rows } = await this.pool.query('SELECT * FROM users WHERE email = $1', [email]);
    return rows[0] ? this.row(rows[0]) : null;
  }
  async findById(id: string) {
    if (!isUuid(id)) return null;
    const { rows } = await this.pool.query('SELECT * FROM users WHERE id = $1', [id]);
    return rows[0] ? this.row(rows[0]) : null;
  }
}

export class EventsRepo implements EventRepository {
  constructor(private pool: Pool) {}
  private row = (r: any): Event => ({
    id: r.id,
    name: r.name,
    venue: r.venue,
    startsAt: r.starts_at,
    salesOpenAt: r.sales_open_at,
    status: r.status as EventStatus,
  });
  async list(cursor: string, limit: number) {
    const cols = 'id, name, venue, starts_at, sales_open_at, status';
    let rows;
    if (!cursor) {
      ({ rows } = await this.pool.query(`SELECT ${cols} FROM events ORDER BY id LIMIT $1`, [limit + 1]));
    } else {
      // The cursor is the last seen id, so it must be a uuid. A malformed cursor is a
      // client error (400), not a reason to hand Postgres bad input.
      if (!isUuid(cursor)) throw Errors.BadRequest;
      ({ rows } = await this.pool.query(
        `SELECT ${cols} FROM events WHERE id > $1 ORDER BY id LIMIT $2`,
        [cursor, limit + 1],
      ));
    }
    const events = rows.map(this.row);
    let nextCursor = '';
    if (events.length > limit) {
      nextCursor = events[limit - 1].id;
      events.length = limit;
    }
    return { events, nextCursor };
  }
  async findById(id: string) {
    if (!isUuid(id)) return null;
    const { rows } = await this.pool.query(
      'SELECT id, name, venue, starts_at, sales_open_at, status FROM events WHERE id = $1',
      [id],
    );
    return rows[0] ? this.row(rows[0]) : null;
  }
}

export class SectorsRepo implements SectorRepository {
  constructor(private pool: Pool) {}
  private row = (r: any): Sector => ({
    id: r.id,
    eventId: r.event_id,
    name: r.name,
    priceCents: Number(r.price_cents),
    currency: r.currency,
    totalInventory: r.total_inventory,
    availableInventory: r.available_inventory,
  });
  async listByEvent(eventId: string) {
    if (!isUuid(eventId)) return [];
    const { rows } = await this.pool.query(
      'SELECT id, event_id, name, price_cents, currency, total_inventory, available_inventory FROM sectors WHERE event_id = $1 ORDER BY name',
      [eventId],
    );
    return rows.map(this.row);
  }
  async findById(id: string) {
    if (!isUuid(id)) return null;
    const { rows } = await this.pool.query(
      'SELECT id, event_id, name, price_cents, currency, total_inventory, available_inventory FROM sectors WHERE id = $1',
      [id],
    );
    return rows[0] ? this.row(rows[0]) : null;
  }
  async decrementInventory(sectorId: string, qty: number) {
    // The anti-overselling primitive. rowCount tells us if the row matched without a
    // separate read; no match means not enough remained.
    const res = await this.pool.query(
      'UPDATE sectors SET available_inventory = available_inventory - $2 WHERE id = $1 AND available_inventory >= $2',
      [sectorId, qty],
    );
    return res.rowCount === 1;
  }
  async incrementInventory(sectorId: string, qty: number) {
    await this.pool.query('UPDATE sectors SET available_inventory = available_inventory + $2 WHERE id = $1', [
      sectorId,
      qty,
    ]);
  }
}

export class QueueRepo implements QueueRepository {
  constructor(private pool: Pool) {}
  async upsert(t: QueueToken) {
    await this.pool.query(
      `INSERT INTO queue_tokens (id, user_id, event_id, position, status, admitted_at)
       VALUES ($1, $2, $3, $4, $5, $6)
       ON CONFLICT (user_id, event_id) DO UPDATE SET status = EXCLUDED.status, admitted_at = EXCLUDED.admitted_at`,
      [t.id, t.userId, t.eventId, t.position, t.status, t.admittedAt],
    );
  }
  async find(userId: string, eventId: string) {
    if (!isUuid(userId) || !isUuid(eventId)) return null;
    const { rows } = await this.pool.query(
      'SELECT id, user_id, event_id, position, status, admitted_at FROM queue_tokens WHERE user_id = $1 AND event_id = $2',
      [userId, eventId],
    );
    const r = rows[0];
    if (!r) return null;
    return {
      id: r.id,
      userId: r.user_id,
      eventId: r.event_id,
      position: r.position,
      status: r.status as QueueStatus,
      admittedAt: r.admitted_at,
    };
  }
  async nextPosition(eventId: string) {
    const { rows } = await this.pool.query(
      'SELECT COALESCE(MAX(position)+1, 0) AS pos FROM queue_tokens WHERE event_id = $1',
      [eventId],
    );
    return Number(rows[0].pos);
  }
}

export class ReservationsRepo implements ReservationRepository {
  constructor(private pool: Pool) {}
  private static COLS =
    'id, user_id, sector_id, quantity, status, expires_at, idempotency_key, created_at';
  private row = (r: any): Reservation => ({
    id: r.id,
    userId: r.user_id,
    sectorId: r.sector_id,
    quantity: r.quantity,
    status: r.status as ReservationStatus,
    expiresAt: r.expires_at,
    idempotencyKey: r.idempotency_key,
    createdAt: r.created_at,
  });
  async create(r: Reservation) {
    try {
      await this.pool.query(
        `INSERT INTO reservations (id, user_id, sector_id, quantity, status, expires_at, idempotency_key, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        [r.id, r.userId, r.sectorId, r.quantity, r.status, r.expiresAt, r.idempotencyKey, r.createdAt],
      );
    } catch (err) {
      if (isUniqueError(err)) throw Errors.Conflict;
      throw err;
    }
  }
  async findById(id: string) {
    if (!isUuid(id)) return null;
    const { rows } = await this.pool.query(`SELECT ${ReservationsRepo.COLS} FROM reservations WHERE id = $1`, [id]);
    return rows[0] ? this.row(rows[0]) : null;
  }
  async findByIdempotencyKey(userId: string, key: string) {
    const { rows } = await this.pool.query(
      `SELECT ${ReservationsRepo.COLS} FROM reservations WHERE user_id = $1 AND idempotency_key = $2`,
      [userId, key],
    );
    return rows[0] ? this.row(rows[0]) : null;
  }
  async updateStatus(id: string, status: string) {
    await this.pool.query('UPDATE reservations SET status = $2 WHERE id = $1', [id, status]);
  }
  async findExpired(now: Date, limit: number) {
    const { rows } = await this.pool.query(
      `SELECT ${ReservationsRepo.COLS} FROM reservations WHERE status = 'held' AND expires_at < $1 LIMIT $2`,
      [now, limit],
    );
    return rows.map(this.row);
  }
}

export class OrdersRepo implements OrderRepository {
  constructor(private pool: Pool) {}
  private static COLS = 'id, reservation_id, user_id, amount_cents, status, idempotency_key, created_at';
  private row = (r: any): Order => ({
    id: r.id,
    reservationId: r.reservation_id,
    userId: r.user_id,
    amountCents: Number(r.amount_cents),
    status: r.status as OrderStatus,
    idempotencyKey: r.idempotency_key,
    createdAt: r.created_at,
  });
  async create(o: Order) {
    try {
      await this.pool.query(
        `INSERT INTO orders (id, reservation_id, user_id, amount_cents, status, idempotency_key, created_at)
         VALUES ($1, $2, $3, $4, $5, $6, now())`,
        [o.id, o.reservationId, o.userId, o.amountCents, o.status, o.idempotencyKey],
      );
    } catch (err) {
      if (isUniqueError(err)) throw Errors.Conflict;
      throw err;
    }
  }
  async findById(id: string) {
    if (!isUuid(id)) return null;
    const { rows } = await this.pool.query(`SELECT ${OrdersRepo.COLS} FROM orders WHERE id = $1`, [id]);
    return rows[0] ? this.row(rows[0]) : null;
  }
  async findByReservationId(reservationId: string) {
    const { rows } = await this.pool.query(`SELECT ${OrdersRepo.COLS} FROM orders WHERE reservation_id = $1`, [
      reservationId,
    ]);
    return rows[0] ? this.row(rows[0]) : null;
  }
  async findByIdempotencyKey(userId: string, key: string) {
    const { rows } = await this.pool.query(
      `SELECT ${OrdersRepo.COLS} FROM orders WHERE user_id = $1 AND idempotency_key = $2`,
      [userId, key],
    );
    return rows[0] ? this.row(rows[0]) : null;
  }
  async updateStatus(id: string, status: string) {
    await this.pool.query('UPDATE orders SET status = $2 WHERE id = $1', [id, status]);
  }
}

export class PaymentsRepo implements PaymentRepository {
  constructor(private pool: Pool) {}
  async upsert(p: Payment) {
    await this.pool.query(
      `INSERT INTO payments (id, order_id, provider_ref, status, attempts)
       VALUES ($1, $2, $3, $4, $5)
       ON CONFLICT (provider_ref) DO UPDATE SET status = EXCLUDED.status, attempts = payments.attempts + 1`,
      [p.id, p.orderId, p.providerRef, p.status, p.attempts],
    );
  }
  async findByOrderId(orderId: string) {
    const { rows } = await this.pool.query(
      'SELECT id, order_id, provider_ref, status, attempts FROM payments WHERE order_id = $1 LIMIT 1',
      [orderId],
    );
    const r = rows[0];
    if (!r) return null;
    return {
      id: r.id,
      orderId: r.order_id,
      providerRef: r.provider_ref,
      status: r.status as PaymentStatus,
      attempts: r.attempts,
    };
  }
}

File: backends/fastapi/app/adapter/postgres.py
"""Postgres adapter over asyncpg. The interesting method is Sectors.decrement_inventory:
a single conditional UPDATE that makes overselling impossible at the database, exactly
as in the Go backend. Only the driver and the language differ.
"""
from __future__ import annotations

import uuid
from datetime import datetime

import asyncpg

from app.domain import errors
from app.domain.models import (
    Event,
    EventStatus,
    Order,
    OrderStatus,
    Payment,
    PaymentStatus,
    QueueStatus,
    QueueToken,
    Reservation,
    ReservationStatus,
    Role,
    Sector,
    User,
)


def _valid_uuid(s: str) -> bool:
    try:
        uuid.UUID(s)
        return True
    except (ValueError, AttributeError, TypeError):
        return False


class Users:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    @staticmethod
    def _row(r) -> User:
        return User(str(r["id"]), r["email"], r["password_hash"], Role(r["role"]), r["created_at"])

    async def find_by_email(self, email: str) -> User | None:
        r = await self._pool.fetchrow(
            "SELECT id, email, password_hash, role, created_at FROM users WHERE email = $1", email
        )
        return self._row(r) if r else None

    async def find_by_id(self, user_id: str) -> User | None:
        if not _valid_uuid(user_id):
            return None
        r = await self._pool.fetchrow(
            "SELECT id, email, password_hash, role, created_at FROM users WHERE id = $1::uuid", user_id
        )
        return self._row(r) if r else None


class Events:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    @staticmethod
    def _row(r) -> Event:
        return Event(str(r["id"]), r["name"], r["venue"], r["starts_at"], r["sales_open_at"], EventStatus(r["status"]))

    async def list(self, cursor: str, limit: int) -> tuple[list[Event], str]:
        cols = "id, name, venue, starts_at, sales_open_at, status"
        if not cursor:
            rows = await self._pool.fetch(f"SELECT {cols} FROM events ORDER BY id LIMIT $1", limit + 1)
        else:
            # The cursor is the last seen id, so it must be a uuid. A malformed cursor
            # is a client error (400), not a reason to hand Postgres bad input.
            if not _valid_uuid(cursor):
                raise errors.BAD_REQUEST
            rows = await self._pool.fetch(
                f"SELECT {cols} FROM events WHERE id > $1::uuid ORDER BY id LIMIT $2", cursor, limit + 1
            )
        events = [self._row(r) for r in rows]
        next_cursor = ""
        if len(events) > limit:
            next_cursor = events[limit - 1].id
            events = events[:limit]
        return events, next_cursor

    async def find_by_id(self, event_id: str) -> Event | None:
        if not _valid_uuid(event_id):
            return None
        r = await self._pool.fetchrow(
            "SELECT id, name, venue, starts_at, sales_open_at, status FROM events WHERE id = $1::uuid", event_id
        )
        return self._row(r) if r else None


class Sectors:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    @staticmethod
    def _row(r) -> Sector:
        return Sector(
            str(r["id"]), str(r["event_id"]), r["name"], r["price_cents"], r["currency"],
            r["total_inventory"], r["available_inventory"],
        )

    async def list_by_event(self, event_id: str) -> list[Sector]:
        if not _valid_uuid(event_id):
            return []
        rows = await self._pool.fetch(
            "SELECT id, event_id, name, price_cents, currency, total_inventory, available_inventory "
            "FROM sectors WHERE event_id = $1::uuid ORDER BY name", event_id
        )
        return [self._row(r) for r in rows]

    async def find_by_id(self, sector_id: str) -> Sector | None:
        if not _valid_uuid(sector_id):
            return None
        r = await self._pool.fetchrow(
            "SELECT id, event_id, name, price_cents, currency, total_inventory, available_inventory "
            "FROM sectors WHERE id = $1::uuid", sector_id
        )
        return self._row(r) if r else None

    async def decrement_inventory(self, sector_id: str, qty: int) -> bool:
        # The anti-overselling primitive. RETURNING tells us if the row matched without
        # a separate read. No match means not enough remained.
        r = await self._pool.fetchrow(
            "UPDATE sectors SET available_inventory = available_inventory - $2 "
            "WHERE id = $1::uuid AND available_inventory >= $2 RETURNING id", sector_id, qty
        )
        return r is not None

    async def increment_inventory(self, sector_id: str, qty: int) -> None:
        await self._pool.execute(
            "UPDATE sectors SET available_inventory = available_inventory + $2 WHERE id = $1::uuid", sector_id, qty
        )


class QueueRepo:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    async def upsert(self, t: QueueToken) -> None:
        await self._pool.execute(
            "INSERT INTO queue_tokens (id, user_id, event_id, position, status, admitted_at) "
            "VALUES ($1::uuid, $2::uuid, $3::uuid, $4, $5, $6) "
            "ON CONFLICT (user_id, event_id) DO UPDATE SET status = EXCLUDED.status, admitted_at = EXCLUDED.admitted_at",
            t.id, t.user_id, t.event_id, t.position, t.status.value, t.admitted_at,
        )

    async def find(self, user_id: str, event_id: str) -> QueueToken | None:
        if not (_valid_uuid(user_id) and _valid_uuid(event_id)):
            return None
        r = await self._pool.fetchrow(
            "SELECT id, user_id, event_id, position, status, admitted_at FROM queue_tokens "
            "WHERE user_id = $1::uuid AND event_id = $2::uuid", user_id, event_id
        )
        if not r:
            return None
        return QueueToken(
            str(r["id"]), str(r["user_id"]), str(r["event_id"]), r["position"],
            QueueStatus(r["status"]), r["admitted_at"],
        )

    async def next_position(self, event_id: str) -> int:
        return await self._pool.fetchval(
            "SELECT COALESCE(MAX(position)+1, 0) FROM queue_tokens WHERE event_id = $1::uuid", event_id
        )


class ReservationRepo:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    @staticmethod
    def _row(r) -> Reservation:
        return Reservation(
            str(r["id"]), str(r["user_id"]), str(r["sector_id"]), r["quantity"],
            ReservationStatus(r["status"]), r["expires_at"], r["idempotency_key"], r["created_at"],
        )

    _COLS = "id, user_id, sector_id, quantity, status, expires_at, idempotency_key, created_at"

    async def create(self, r: Reservation) -> None:
        try:
            await self._pool.execute(
                "INSERT INTO reservations (id, user_id, sector_id, quantity, status, expires_at, idempotency_key, created_at) "
                "VALUES ($1::uuid, $2::uuid, $3::uuid, $4, $5, $6, $7, $8)",
                r.id, r.user_id, r.sector_id, r.quantity, r.status.value, r.expires_at, r.idempotency_key, r.created_at,
            )
        except asyncpg.UniqueViolationError as exc:
            raise errors.CONFLICT from exc

    async def find_by_id(self, reservation_id: str) -> Reservation | None:
        if not _valid_uuid(reservation_id):
            return None
        r = await self._pool.fetchrow(f"SELECT {self._COLS} FROM reservations WHERE id = $1::uuid", reservation_id)
        return self._row(r) if r else None

    async def find_by_idempotency_key(self, user_id: str, key: str) -> Reservation | None:
        r = await self._pool.fetchrow(
            f"SELECT {self._COLS} FROM reservations WHERE user_id = $1::uuid AND idempotency_key = $2", user_id, key
        )
        return self._row(r) if r else None

    async def update_status(self, reservation_id: str, status: str) -> None:
        await self._pool.execute("UPDATE reservations SET status = $2 WHERE id = $1::uuid", reservation_id, status)

    async def find_expired(self, now: datetime, limit: int) -> list[Reservation]:
        rows = await self._pool.fetch(
            f"SELECT {self._COLS} FROM reservations WHERE status = 'held' AND expires_at < $1 LIMIT $2", now, limit
        )
        return [self._row(r) for r in rows]


class OrderRepo:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    _COLS = "id, reservation_id, user_id, amount_cents, status, idempotency_key, created_at"

    @staticmethod
    def _row(r) -> Order:
        return Order(
            str(r["id"]), str(r["reservation_id"]), str(r["user_id"]), r["amount_cents"],
            OrderStatus(r["status"]), r["idempotency_key"], r["created_at"],
        )

    async def create(self, o: Order) -> None:
        try:
            await self._pool.execute(
                "INSERT INTO orders (id, reservation_id, user_id, amount_cents, status, idempotency_key, created_at) "
                "VALUES ($1::uuid, $2::uuid, $3::uuid, $4, $5, $6, now())",
                o.id, o.reservation_id, o.user_id, o.amount_cents, o.status.value, o.idempotency_key,
            )
        except asyncpg.UniqueViolationError as exc:
            raise errors.CONFLICT from exc

    async def find_by_id(self, order_id: str) -> Order | None:
        if not _valid_uuid(order_id):
            return None
        r = await self._pool.fetchrow(f"SELECT {self._COLS} FROM orders WHERE id = $1::uuid", order_id)
        return self._row(r) if r else None

    async def find_by_reservation_id(self, reservation_id: str) -> Order | None:
        r = await self._pool.fetchrow(f"SELECT {self._COLS} FROM orders WHERE reservation_id = $1::uuid", reservation_id)
        return self._row(r) if r else None

    async def find_by_idempotency_key(self, user_id: str, key: str) -> Order | None:
        r = await self._pool.fetchrow(
            f"SELECT {self._COLS} FROM orders WHERE user_id = $1::uuid AND idempotency_key = $2", user_id, key
        )
        return self._row(r) if r else None

    async def update_status(self, order_id: str, status: str) -> None:
        await self._pool.execute("UPDATE orders SET status = $2 WHERE id = $1::uuid", order_id, status)


class PaymentRepo:
    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    async def upsert(self, p: Payment) -> None:
        await self._pool.execute(
            "INSERT INTO payments (id, order_id, provider_ref, status, attempts) "
            "VALUES ($1::uuid, $2::uuid, $3, $4, $5) "
            "ON CONFLICT (provider_ref) DO UPDATE SET status = EXCLUDED.status, attempts = payments.attempts + 1",
            p.id, p.order_id, p.provider_ref, p.status.value, p.attempts,
        )

    async def find_by_order_id(self, order_id: str) -> Payment | None:
        r = await self._pool.fetchrow(
            "SELECT id, order_id, provider_ref, status, attempts FROM payments WHERE order_id = $1::uuid LIMIT 1", order_id
        )
        if not r:
            return None
        return Payment(str(r["id"]), str(r["order_id"]), r["provider_ref"], PaymentStatus(r["status"]), r["attempts"])
